_posts/2019-10-10-anonymous-access-and-social-login.adoc

```diff
@@ -0,0 +1,1275 @@
+
+    public static final String NAME = "petclinic_SocialLoginProvider";
+
+    @Inject
+    private SocialRegistrationService socialRegistrationService;
+
+    @Nullable
+    @Override
+    public AuthenticationDetails login(Credentials credentials) throws LoginException {
+        SocialCredentials socialCredentials = (SocialCredentials) credentials;
+
+        SocialLoginService.SocialUserData userData = socialCredentials.getUserData();
+
+        // to be implemented;
+        User user = null;
+
+        Locale defaultLocale = socialCredentials.getLocale();
+
+        return super.login(new ExternalUserCredentials(user.getLogin(), defaultLocale));
+    }
+
+    // ...
+}
+----
+
+To form credentials we have to find an existing or register a new user.
+
+=== User Registration
+
+Create a new entity `SocialUser` extending `User` and add three fields to it:
+
+* `googleId`
+* `facebookId`
+* `githubId`
+
+These fields are required to bind the social network profile with the system user to be able to find it later. We should also assign a default group to a new user - create a new configuration interface:
+
+.SocialRegistrationConfig.java
+[source, java]
+----
+@Source(type = SourceType.APP)
+public interface SocialRegistrationConfig extends Config {
+
+    @Default("0fa2b1a5-1d68-4d69-9fbd-dff348347f93")
+    @Property("social.defaultGroupId")
+    @Factory(factory = UuidTypeFactory.class)
+    UUID getDefaultGroupId();
+}
+----
+
+Create a new service `SocialRegistrationService` intended to find existing or register a new user:
+
+.SocialRegistrationService.java
+[source, java]
+----
+public interface SocialRegistrationService {
+
+    String NAME = "petclinic_SocialRegistrationService";
+
+    User findOrRegisterUser(String socialServiceId, String login, String name,
+                            SocialService socialService);
+
+    // ...
+}
+----
+
+Its implementation is quite straightforward:
+
+.SocialRegistrationServiceBean.java
+[source, java]
+----
+public class SocialRegistrationServiceBean implements SocialRegistrationService {
+
+    private static final Pattern EMAIL_PATTERN = Pattern.compile("[^@]+@[^.]+\\..+");
+
+    @Inject
+    private DataManager dataManager;
+    @Inject
+    private Configuration configuration;
+
+    @Override
+    public User findOrRegisterUser(String socialServiceId, String login, String name,
+                                   SocialService socialService) {
+        User existingUser = findExistingUser(socialService, socialServiceId);
+        if (existingUser != null) {
+            return existingUser;
+        }
+
+        SocialUser user = createNewUser(socialServiceId, login, name, socialService);
+
+        return dataManager.commit(user);
+    }
+
+    @Nullable
+    private User findExistingUser(SocialService socialService, String socialServiceId) {
+        String socialServiceField = getSocialIdParamName(socialService);
+
+        LoadContext<User> ctx = new LoadContext<>(User.class);
```

Comment:
Let's promote the modern way: `dataManager.load(User.class).query("...").parameter(...).one()`.
And you don't have to specify local view, it is by default.
Fix `getDefaultGroup()` below too.